lunatrace/bsl/backend/src/github/actions/generate-snapshots-for-repositories.ts

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright by LunaSec (owned by Refinery Labs, Inc)
+ *
+ * Licensed under the Business Source License v1.1 
+ * (the "License"); you may not use this file except in compliance with the
+ * License. You may obtain a copy of the License at
+ *
+ * https://github.com/lunasec-io/lunasec/blob/master/licenses/BSL-LunaTrace.txt
+ *
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+import { GithubRepositoryInfo } from "../../types/github";
+import {log} from "../../utils/log";
+import {catchError, threwError} from "../../utils/try";
+import {generateGithubGraphqlClient} from "../api";
+import {getInstallationAccessToken} from "../auth";
+
+import {generateSnapshotForRepository} from "./generate-snapshot-for-repository";
+
+export async function generateSnapshotsForGithubRepos(installationId: number, githubRepos: GithubRepositoryInfo[]) {
+  const installationToken = await getInstallationAccessToken(installationId);
+
+  if (installationToken.error) {
+    log.error('unable to get installation token', {
+      error: installationToken.msg
+    });
+    return;
+  }
+
+  const github = generateGithubGraphqlClient(installationToken.res);
+
+  await Promise.all(githubRepos.map(async (repo) => {
+    const githubRepo = await catchError(async () => await github.GetRepository({
+      name: repo.repoName,
+      owner: repo.orgName
+    }));
+
+    if (threwError(githubRepo)) {
+      log.error('unable to get github repo information', {
+        repo,
+        msg: githubRepo.message
+      });
+      return;
+    }
+
+    if (!githubRepo.repository || !githubRepo.repository.defaultBranchRef) {
+      log.error('github repository information is not defined', {
+        repo,
+        githubRepo
+      });
+      return;
+    }
+
+    const cloneUrl = `${githubRepo.repository.url}.git`;
+    const gitBranch = githubRepo.repository.defaultBranchRef.name;
+
+    await generateSnapshotForRepository({
```

Comment:
Just to confirm, does this do a shallow Git clone? I see that it calls into the CLI, but I just want to verify.

It might not always be possible. I thiiiiink you probably can't always have a shallow clone because, iirc, you can't "check out a given commit hash" when you've done a shallow clone. If that's true, that means that there is a race condition (you try to checkout the commit that was the "head" but then another push happens while this job is running).

Maybe the workaround then is to just use `--single-branch`? I'm punting this to you to verify though. :pray: